pymc3/backends/arviz.py

```diff
@@ -0,0 +1,606 @@
+
+        self.attrs = None
+        if trace is not None:
+            self.nchains = trace.nchains if hasattr(trace, "nchains") else 1
+            if hasattr(trace.report, "n_draws") and trace.report.n_draws is not None:
+                self.ndraws = trace.report.n_draws
+                self.attrs = {
+                    "sampling_time": trace.report.t_sampling,
+                    "tuning_steps": trace.report.n_tune,
+                }
+            else:
+                self.ndraws = len(trace)
+                if self.save_warmup:
+                    warnings.warn(
+                        "Warmup samples will be stored in posterior group and will not be"
+                        " excluded from stats and diagnostics."
+                        " Please consider using PyMC3>=3.9 and do not slice the trace manually.",
+                        UserWarning,
+                    )
+            self.ntune = len(self.trace) - self.ndraws
+            self.posterior_trace, self.warmup_trace = self.split_trace()
+        else:
+            self.nchains = self.ndraws = 0
+
+        self.prior = prior
+        self.posterior_predictive = posterior_predictive
+        self.log_likelihood = log_likelihood
+        self.predictions = predictions
+        self.index_origin = rcParams["data.index_origin"] if index_origin is None else index_origin
+
+        def arbitrary_element(dct: Dict[Any, np.ndarray]) -> np.ndarray:
+            return next(iter(dct.values()))
+
+        if trace is None:
+            # if you have a posterior_predictive built with keep_dims,
+            # you'll lose here, but there's nothing I can do about that.
+            self.nchains = 1
+            get_from = None
+            if predictions is not None:
+                get_from = predictions
+            elif posterior_predictive is not None:
+                get_from = posterior_predictive
+            elif prior is not None:
+                get_from = prior
+            if get_from is None:
+                # pylint: disable=line-too-long
+                raise ValueError(
+                    "When constructing InferenceData must have at least"
+                    " one of trace, prior, posterior_predictive or predictions."
+                )
+
+            aelem = arbitrary_element(get_from)
+            self.ndraws = aelem.shape[0]
+
+        self.coords = {} if coords is None else coords
+        if hasattr(self.model, "coords"):
+            self.coords = {**self.model.coords, **self.coords}
+
+        self.dims = {} if dims is None else dims
+        if hasattr(self.model, "RV_dims"):
+            model_dims = {k: list(v) for k, v in self.model.RV_dims.items()}
+            self.dims = {**model_dims, **self.dims}
+
+        self.density_dist_obs = density_dist_obs
+        self.observations, self.multi_observations = self.find_observations()
+
+    def find_observations(self) -> Tuple[Optional[Dict[str, Var]], Optional[Dict[str, Var]]]:
```

Comment:
This probably has to be rewritten from scratch. Do `MultiObservedRV` still exist?

Reply:
No, that class is gone, but there will likely be a way to identify a variable as one that _would have_ been a `MultiObservedRV` in the past.  

Currently, `MultiObservedRV` is one of those cases that I haven't covered in https://github.com/pymc-devs/pymc3/issues/4463.  If you know how they work, tell me and I'll port the functionality into `v4`.

Reply:
I am not sure if MultiObservedRV was used by something other than DensityDist, but just in case, here is what I know about them. 

The main functionality of `MultiObservedRV` was to allow passing multiple arrays (with potentially different types, shapes or whatever) as `observed` kwarg using a dictionary. It then calculated the log likelihood values from this dict instead of a single tensor like ObservedRV. 

The second important difference was that (unlike `ObservedRV`) `MultiObservedRV` did not check that the values in the dict passed as `observed` were not other `ObservedRV` nor `FreeRV`, which is what ended up motivating the `given` argument to try and sort this out. 

Minor difference is that MultiObservedRV stored these dict observations as the attribute `data` instead of using `observations`. 

What I know nothing about are the motivations behind any of these behaviours. I progressively discovered that while trying to integrate MultiObservedRV with ArviZ.

Reply:
Thanks, those details help a lot.  I'm still a bit confused about how these different observations in such a `dict` map to the underlying statistics/math, though.  Is each item in the `dict` a different copy of the underlying random variable (e.g. `Normal(0, 1, observed=multi_obs_dict)` maps to `[Normal(0, 1, observed=val) for val in multi_obs_dict.values()]`?

I don't plan on re-implementing the `MultiObservedRV`-like functionality any time soon, but, if someone else wants to give it a shot, I'll help where I can.

Reply:
I don't think so, the dict was interpreted as kwargs for the logp calculation. One example of a weighted normal via DensityDist (taken from https://github.com/arviz-devs/arviz/issues/384) would be:

```
with pm.Model():
    a = pm.Normal('a', 0, 10)
    b = pm.Normal('b', 0, 10)
    
    mu = a + b*X
    sigma = pm.HalfNormal('sigma', 1)
    
    def weighted_normal(y, w):
        return w*pm.Normal.dist(mu=mu, sd=sigma).logp(y)
    
    y_logp = pm.DensityDist('y_logp', weighted_normal, observed={'y': Y, 'w': W})
```

It is problematic because there are several issues that I don't really understand nor even distinguish properly, I think that MultiObservedRV exists only to be used via densitydist, and I think that everything that one can achive with DensityDist can be done also with potentials, but I am not sure of either of the two.

Reply:
Oh, wow.  Yeah, I don't think we want to do that at all; it appears to break the meaning of the `observed` parameter severely.

Reply:
Yes, and what is even worse is that not only it accepts `y` and `w`, it would still work if `mu` and `sigma` were passed as `observed` when using the dict. Sampling somehow works on the PyMC3 side, but it breaks ArviZ

Reply:
I have never used `MultiObservedRV`, but in connection to your question, @OriolAbril, about the motivation, was it for Variational Inference? Or am I confusing it with `MiniBatch` (which I also don't really understand)?